tests/test_rp_exec.py

```diff
@@ -85,28 +86,53 @@ async def test_raptor_master(pilot_description, rp_venv, cleandir):
                     "scheduler": scheduler.uid,
                     "mode": scalems.radical.raptor.CPI_MESSAGE,
                     "metadata": scalems.messages.HelloCommand().encode(),
+                    "uid": "task-hello",
                 }
             )
             logger.debug(f"Submitting {str(td.as_dict())}")
             tasks = await to_thread(dispatcher.runtime.task_manager().submit_tasks, [td])
             hello_task = tasks[0]
             logger.debug(f"Submitted {str(hello_task.as_dict())}. Waiting...")
-            state = await to_thread(hello_task.wait, state=rp.FINAL, timeout=180)
+            state = await to_thread(hello_task.wait, state=rp.FINAL, timeout=timeout)
             logger.debug(str(hello_task.as_dict()))
 
             td.metadata = scalems.messages.StopCommand().encode()
+            td.output_staging = []
+            td.uid = "task-stop"
             logger.debug(f"Submitting {str(td.as_dict())}")
             tasks = await to_thread(dispatcher.runtime.task_manager().submit_tasks, [td])
             stop_task = tasks[0]
 
-            # Note: Once `stop` is issued, the client will never see the Task complete. I.e. the following would fail:
-            # stop_watcher = asyncio.create_task(
-            #     to_thread(stop_task.wait, state=rp.FINAL, timeout=180), name='stop-watcher')
-            # await asyncio.wait_for(stop_watcher, timeout=120)
-            assert stop_task.state not in rp.FINAL
+            # We expect the status update -> DONE, even if self.stop() was called during result_cb for the task.
+            stop_watcher = asyncio.create_task(
+                to_thread(stop_task.wait, state=rp.FINAL, timeout=timeout), name="stop-watcher"
+            )
+
+            scheduler_watcher = asyncio.create_task(
+                to_thread(scheduler.wait, state=rp.FINAL, timeout=timeout), name="master-watcher"
+            )
+            # If master task fails, stop-watcher will never complete.
+            done, pending = await asyncio.wait(
+                (stop_watcher, scheduler_watcher), timeout=timeout, return_when=asyncio.FIRST_COMPLETED
+            )
 
-            # Note: We don't actually have anything to keep us from canceling the Master task
-            # before the work has been handled.
+            if scheduler_watcher not in done:
+                await asyncio.wait_for(scheduler_watcher, timeout=10)
+            logger.debug(f"scheduler-task state: {scheduler.state}")
+            if scheduler.state == rp.DONE and stop_watcher in pending:
```

Comment:
maybe `if scheduler.state in rp.FINAL and ...`?

Reply:
I specifically want to catch the case where the scheduler appeared to finish successfully before the stop task. scheduler finishing unsuccessfully is handled as a distinct logical condition.